### ISSUE TITLE
Impossible to get the first day of any 2016 week with other locale than english

### ISSUE THREAD

**Issue by Nicofuma** [moment/moment#3696]
**Description of the Issue and Steps to Reproduce:**
```
console.log('2016 - 20 === ', moment().locale('fr').year(2016).week(20).startOf('week').format('dddd Do MMMM YYYY')); // 2016 - 20 ===  lundi 11 mai 2015
console.log('2016 - 20 === ', moment().locale('en').year(2016).week(20).startOf('week').format('dddd Do MMMM YYYY')); // 2016 - 20 ===  Sunday 8th May 2016
console.log('2015 - 20 === ', moment().locale('fr').year(2015).week(20).startOf('week').format('dddd Do MMMM YYYY')); // 2016 - 20 ===  lundi 11 mai 2015
```

I'm expecting the first line to give a date in 2016 (like the second one), not 2015.
I don't know if I am using the library the wrong way or if it is a bug, but I have the same issue with every week, it looks like the 2016 year does not exist.

I also reproduced the issue with other locales like `it` or `de`.

**Environment:**

Chrome 55 on Ubuntu 16.0.4 64 bits. I run my tests using the javascript console on my browser on https://momentjs.com

If you are reporting an issue, please run the following code in the environment you are using and include the output:
```js
Mon Jan 02 2017 16:34:31 GMT+0100 (CET)
02/01/2017 à 16:34:31
-60
Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/55.0.2883.87 Safari/537.36
2.17.1
```

**Comment by clemherreman** [comment]
Same error here, same output

**Comment by hbierlee** [comment]
Same problems here. It's setting of the week that seems to mess it up. Using `nl` locale:

```
var date = moment().year(2016).week(50);
console.log(date);  // moment("2015-12-12T20:30:08.380")
```

Without setting the week:
```
var date = moment().year(2016); 
console.log(date);  // moment("2016-01-02T20:31:26.205")
```

**Comment by angusterravion** [comment]
moment().locale('en', null).millisecond(0).second(0).minute(0).hour(0).year(2016).isoWeek(48).isoWeekday(1);

Gives me:
Mon Nov 23 2015 00:00:00 GMT-0800 (PST)

**Comment by dmaicher** [comment]
Same problem here: 

```html
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <title>Title</title>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/moment.js/2.17.1/moment.min.js"></script>
</head>
<body>
    <script>
        console.log(moment().year(2015).isoWeek(40).format('YYYY MM DD'));
        console.log(moment().year(2016).isoWeek(40).format('YYYY MM DD'));
        console.log(moment().year(2017).isoWeek(40).format('YYYY MM DD'));
    </script>
</body>
</html>
```

Gives me:

```
2015 10 03
2015 10 04
2017 10 03
```

This is pretty urgent for us :cry:

**Comment by ar2rsawseen** [comment]
This is because you are changing year and not date, so

```
var m = moment();  //creates current date object
m.year(2016); //sets year to 2016 but date is still January 3rd
m.isoWeek(44) // subtracts 63 days (current week (53 - 44)*7 = 63) which means it is now 2015
```

The solution would be to do
```
var m = moment().year(2016).endOf("year");
console.log(m.isoWeek(44).year());
```

or for momentjs to calculate the date of 44 isoweek for provided year instead of subtracting dats

**Comment by Nicofuma** [comment]
So it is a missuses of the library, thanks. But do you understand why the dates seems to be correct if the locale is english?

**Comment by ar2rsawseen** [comment]
No, actually I think it is bug in the library, trying figure out how to fix it :)

**Comment by ar2rsawseen** [comment]
I am not familiar with internals, but if I am understanding it correctly, then maybe calculating isoweek would be better done by setting day to firstWeekOffset() + input*7

Something like

```
function getSetISOWeek (input) {
    if(input == null){
        return weekOfYear(this, 1, 4).week;
    }
    else{
        set$1(this, 'Date', firstWeekOffset(this.year(), 1, 4)+input*7);
        return this;
    }
}
```

**Comment by Ikkyusan1** [comment]
I'm encoutering the same problem too. Your workaround did the trick though : 
```
let failing = moment.utc().year(2016).week(1).weekday(0).startOf('day');
console.log(failing.format('YYYY-MM-DD')); // 2014-12-29

let working = moment.utc().endOf('year');
working.year(2016).week(1).weekday(0).startOf('day');
console.log(working.format('YYYY-MM-DD')); // 2016-01-04
```

**Comment by ar2rsawseen** [comment]
It will do the trick to the year, but will fail for the date, I think correct implementation should be accounting for year change when calculating isoweek

```
function getSetISOWeek (input) {
    var woy = weekOfYear(this, 1, 4);
    return input == null ? woy.week : this.add((input - woy.week + (this.year()-woy.year)*52) * 7, 'd');
}
```

**Comment by kiliangrashoff** [comment]
tldr use the following code to get the 1st week of 2016 according to ISO 8601 correctly:
```
moment().isoWeekYear(2016).isoWeek(1)
```
If you always use isoWeekYear() to set a year in which you want to get a isoWeek, and use year() when you want to set a year in a date, the issues stated above do not occur.

For the mentioned locales, the week system is ISO 8601, and therefore moment's week() function is (effectively) the same as the isoWeek() function. All of the following is therefore according to ISO 8601. North-american locales should not have this issue, but some others with weeks that overlap the year boundary may.

Week 01 of a given year [is the week in which four or more days are in that year](https://en.wikipedia.org/wiki/ISO_week_date#First_week). The year that the ISO week corresponds to is called the week-year, which you can get and set with the isoWeekYear() function in moment. Therefore, the first three days of a year can be in a previous week-year and reversely, the last three days of a year may be in the next week-year. This may be hard to follow, but it is by design, the wiki article gives more explanation.

Where this gets confusing is that isoWeek(), and week() for certain locales, sets the week in the same week-year. So the following is true:
```
date = moment({year:2016, month:0, day:3})
console.log(date.format("YYYY-MM-DD")) //2016-01-03
console.log(date.isoWeekYear()) //2015
```
And any call to isoWeek(), or week() for certain locales, will set that week in 2015. I don't know whether this is by design or a bug, since the documentation is unclear on what year will be used for calls to week(). If this behavior is by design, it may be useful to mention this more clearly in the documentation.

Please note that using endOf('year') is not a good solution, because the last day of the year may be in week 01 of the next year, causing the same issue:
```
console.log(moment({year:2018}).endOf('year').isoWeekYear()) //2019
```

**Comment by icambron** [comment]
Couldn't have said better @kiliangrashoff -- that's absolutely right.

**Comment by ar2rsawseen** [comment]
yes, but it does not make sense for end user to use isoWeekYear if he is using weeks and locale happens to be iso standard. How would I know it is iso standard?
